src/superqt/utils/_ensure_thread.py

```diff
@@ -166,17 +168,23 @@ def _run_in_thread(
     thread: QThread,
     await_return: bool,
     timeout: int,
-    *args,
-    **kwargs,
-):
+    args: tuple,
+    kwargs: dict,
+) -> Any:
+    try:
+        max_args: int | None = func.__code__.co_argcount
+    except AttributeError:
+        max_args = None
```

Comment:
this will not work properly:

```python
from functools import wraps

def func(a, b):
    return a + b

def wr(fun):
    @wraps(fun)
    def wr2(*args):
        return fun(*args) * 2
    return wr2

fn2 = wr(func)
```

then `fn2.__code__.co_argcount` will return 0.

This reveals two problems. At first this approach does not unwrap functions. 
At second the `*args` is not counted:

```python
>> def a(b, *c):
..    pass
>> a.__code__.co_argcount
1
```

Maybe use inspect (or reimplement most of logic, to handle partials etc)

Reply:
thanks, yeah I was trying to avoid using inspect (so heavy...) but you're probably right.  lemme see if I can make a failing test then fix it

Reply:
yeah, I'm doing away with it, going back to inspect... but also moving where the inspection occurs higher up (to avoid doing it often)

Reply:
actually. I think this one is correct by the way.  Signals include `self` in their connections... i think I tested that

Reply:
ok updated... but all I really did was add an unwrap call:
```python
def _get_max_args(func: Callable) -> int | None:
     try:
         return inspect.unwrap(func).__code__.co_argcount  # type: ignore
     except Exception:
         return None
```

I added a number of additional tests, and I can't find a case where this simpler/faster approach fails.  so, if you still have issues with it, can you demonstrate a failure?

Reply:
eh, never mind.  I changed it to:
```python
def _get_max_args(func: Callable) -> int | None:
    try:
        return len(inspect.signature(func).parameters)
    except Exception:
        return None
```

I do think the co_argcount is sufficient, and it's ~6µs instead of 50µs ... but it should only happen once (at decoration time), and i guess it's more likely to catch extreme cases.  so let's just play it safe

Reply:
I think that you should add these tests. It will be helpful if you try to speed up code.  (please notice that the first one is failing for current implementation):

```python
def test_get_max_args_simple():

    def fun1():
        pass

    assert _get_max_args(fun1) == 0

    def fun2(a):
        pass

    assert _get_max_args(fun2) == 1

    def fun3(a, b=1):
        pass

    assert _get_max_args(fun3) == 2

    def fun4(a, *, b=2):
        pass

    assert _get_max_args(fun4) == 1

    def fun5(a, *b):
        pass

    assert _get_max_args(fun5) is None



def test_get_max_args_wrapped():
    from functools import partial, wraps

    def fun1(a, b):
        pass

    assert _get_max_args(partial(fun1, 1)) == 1

    def dec(fun):
        @wraps(fun)
        def wrapper(*args, **kwargs):
            return fun(*args, **kwargs)
        return wrapper

    assert _get_max_args(dec(fun1)) == 2


def test_get_max_args_methods():

    class A:
        def fun1(self):
            pass

        def fun2(self, a):
            pass

        def __call__(self, a, b=1):
            pass


    assert _get_max_args(A().fun1) == 0
    assert _get_max_args(A().fun2) == 1
    assert _get_max_args(A()) == 2
```

Reply:
ok, the following works for these cases:

```python
def get_max_args(func: Callable) -> int | None:
    """Return the maximum number of positional arguments that func can accept."""
    sig = signature(func)
    max_args = 0
    for param in sig.parameters.values():
        if param.kind == param.VAR_POSITIONAL:
            return None
        if param.kind in {param.POSITIONAL_ONLY, param.POSITIONAL_OR_KEYWORD}:
            max_args += 1
    return max_args
```

Reply:
updated to use it here, along with your tests